keras/layers/dense_attention.py

```diff
@@ -40,6 +41,9 @@ class BaseDenseAttention(base_layer.BaseRandomLayer):
       flow of information from the future towards the past.
     dropout: Float between 0 and 1. Fraction of the units to drop for the
       attention scores.
+    return_attention_scores: bool, if `True`, layer returns the attention scores.
+      If `True`, the layer is incompatible with TimeDistributed wrapper.
```

Comment:
If this change is purely trying to address the usage of TimeDistributed layer, then how about just let compute_output_shape() to assume return_attention_scores is False?

Moving this call param to init is a quite big API contract change (even we are just marking it as deprecated), there are many lib down the stream that is also using this as a call arg, and the migration effort to make it a call arg is potentially big.

Reply:
Hm, looks like this is the way to solve the issue; but I'm not sure if it's okay to return tensor(s) with shapes that differ from the shape returned by `compute_output_shape`. It would not cause any problems programmatically, but might cause a problem with interpretation of how the layer works. It would confuse user in some (rare) cases:

```python3
attention_layer = keras.layers.Attention()
result, scores = attention_layer([[query_input, value_input], return_attention_scores=True)
res_output_shape, scores_shape = attention_layer.comute_output_shape()  # a user might expect shapes for both attention tensor and weights, but will get an error;
```
We could overcome this by adding a parameter `return_attention_scores` to the `compute_output_shape`, but it introduces a discrepancy between a method of this layer and the rest of keras layers. It seems to me that it's a bad change.
The second problem I expect if one will try to use returned attention weights as an input to the next layer in a Functional model. `compute_output_shape`  will return a value which is different from the actual output. I've not checked yet, but it looks like it would cause a model compilation error. Unfortunately, it's impossible to provide a `return_attention_scores` directly to the `attention_layer.compute_output_shape` method, because it requires unacceptable amount of (excessive for anything else) changes. 

So that I decided to move `compute_output_shape` to the constructor, realising that this is a significant API change.

Also there's a second reason for this merge request, because I'm not purely addressing the usage of TimeDistributed layer. This parameter was originally introduced in [this](https://github.com/tensorflow/tensorflow/issues/44127) feature request. But it seems to me that the solution wasn't optimal. `BaseDenseAttention.call` method has parameters that change output shape of the method. But for the rest of the layers parameters that could cause such changes are placed in a layer's constructor. And in this request I apply the same 'idea' to the `BaseDenseAttention`.

--
Considering this, I need a guidance if I should add changes you suggested or it's better to close PR (because original issue about TimeDistributed layer was resolved).

Reply:
So, I've made this change: `compute_output_shape` assumes that `return_attention_scores` is False.

Reply:
Thanks, for compute_output_shape(), it is mostly used by the framework that is trying to infer the output shape in some corner case. Most of the time, caller will just use the __call__ method to get the output. So I think this change should be fine.